### ISSUE TITLE
Regression: incorrectly reported TypeDoesNotContainType in a loop

### ISSUE THREAD

**Issue by ArtemGoutsoul** [vimeo/psalm#10716]
Psalm started incorrectly reporting TypeDoesNotContainType some time after [5.19.1](https://github.com/vimeo/psalm/releases/tag/5.19.1)

Consider a two dimensional array, the code returns a count for each unique value in "key" column
```php
/**
 * @param array{key: string}[] $array
 */
function func(array $array): array
{
    $r = [];
    foreach ($array as $row) {
        $r[$row['key']] = empty($r[$row['key']])
            ? 1
            : $r[$row['key']] + 1;
    }
    return $r;
}
```

psalm reports
```
Psalm output (using commit [c488d40](https://github.com/vimeo/psalm/commit/c488d40)): 

ERROR: [TypeDoesNotContainType](https://psalm.dev/056) - 10:27 - Operand of type false is always falsy
```

https://psalm.dev/r/707051b9f6

**Comment by ArtemGoutsoul** [comment]
@kkmuffme interesting!

I guess an earlier issue manifests differently now. But it's definitely a regression specifically for the use case I posted. It worked ok before, but fails now... But yeah, there's clearly some issue with type detection when an array is assigned in a loop.
